src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -772,79 +774,36 @@ ord4 c = assert (n >= 0x10000)  (WCons x1 (WCons x2 (WCons x3 (WCons x4 WNil))))
+readCharUtf8With :: Monad m => WList -> Unfold m Char Word8
+readCharUtf8With surr = Unfold step inject
 
-{-# ANN type InvalidAction Fuse #-}
-data InvalidAction =
-    DropInvalid | ErrorInvalid | IgnoreInvalid | ReplaceInvalid
-
-replaceInvalid :: s -> Step (EncodeState s) a
-replaceInvalid s =
-    Skip $ EncodeState s (WCons 239 (WCons 191 (WCons 189 WNil)))
+    where
 
-dropInvalid :: s -> Step (EncodeState s) a
-dropInvalid s = Skip (EncodeState s WNil)
+    inject c =
+        return $ case ord c of
+            x | x <= 0x7F -> fromIntegral x `WCons` WNil
+              | x <= 0x7FF -> ord2 c
+              | x <= 0xFFFF -> if isSurrogate c then surr else ord3 c
+              | otherwise -> ord4 c
 
-errorOnInvalid :: s -> Step (EncodeState s) a
-errorOnInvalid _ =
-    error $
-    show "Streamly.Internal.Data.Stream.StreamD.encodeUtf8:"
-    ++ "Encountered a surrogate"
+    {-# INLINE_LATE step #-}
+    step WNil = return Stop
+    step (WCons x xs) = return $ Yield x xs
 
-{-# INLINE_NORMAL encodeUtf8DGeneric #-}
-encodeUtf8DGeneric ::
-       Monad m
-    => InvalidAction
-    -> Stream m Char
-    -> Stream m Word8
-encodeUtf8DGeneric act (Stream step state) =
-    Stream step' (EncodeState state WNil)
-
-    where
-
-    {-# INLINE_LATE step' #-}
-    step' gst (EncodeState st WNil) = do
-        r <- step (adaptState gst) st
-        return $
-            case r of
-                Yield c s ->
-                    case ord c of
-                        x | x <= 0x7F ->
-                                Yield (fromIntegral x) (EncodeState s WNil)
-                            | x <= 0x7FF -> Skip (EncodeState s (ord2 c))
-                            | x <= 0xFFFF ->
-                                case act of
-                                    DropInvalid ->
-                                        if isSurrogate c
-                                        then dropInvalid s
-                                        else Skip (EncodeState s (ord3 c))
-
-                                    ErrorInvalid ->
-                                        if isSurrogate c
-                                        then errorOnInvalid s
-                                        else Skip (EncodeState s (ord3 c))
-
-                                    IgnoreInvalid ->
-                                        Skip (EncodeState s (ord3 c))
-
-                                    ReplaceInvalid ->
-                                        if isSurrogate c
-                                        then replaceInvalid s
-                                        else Skip (EncodeState s (ord3 c))
-
-                            | otherwise -> Skip (EncodeState s (ord4 c))
-                Skip s -> Skip (EncodeState s WNil)
-                Stop -> Stop
-    step' _ (EncodeState s (WCons x xs)) = return $ Yield x (EncodeState s xs)
+{-# INLINE_NORMAL readCharUtf8' #-}
+readCharUtf8' :: Monad m => Unfold m Char Word8
```

Comment:
This will be moved to the `Char` module. We can perhaps rename this to just `readUtf8'` or just `read'`.  Need to think about consistency between folds/parsers and unfold naming. The opposite operation to parse a Char from Word8 stream would be `write` or `writeUtf8` or `parse`?

Or would `unfold`, `fold`, `parse` be better names than `read`, `write` etc. That would require a lot of changes to the existing modules though, but we can do it if its worth it and more intuitive. Also, in that case we may have to change the `unfold`, `fold` in stream to something like `runUnfold`, `runFold`.